src/libs/ajax/data-table-providers/WdsDataTableProvider.test.ts

```diff
@@ -0,0 +1,302 @@
+import { Ajax } from 'src/libs/ajax'
+import { asMockedFn } from 'src/test-utils'
+
+import { EntityMetadata, EntityQueryOptions, EntityQueryResponse } from './DataTableProvider'
+import { RecordQueryResponse, RecordTypeSchema, SearchRequest, WdsDataTableProvider, wdsToEntityServiceMetadata } from './WdsDataTableProvider'
+
+
+jest.mock('src/libs/ajax')
+
+const uuid = '123e4567-e89b-12d3-a456-426614174000' // value doesn't matter for these tests
+
+// shell class that extends WdsDataTableProvider to allow testing protected methods
+class TestableWdsProvider extends WdsDataTableProvider {
+  transformPageOverride(wdsPage: RecordQueryResponse, recordType: string, queryOptions: EntityQueryOptions): EntityQueryResponse {
+    return this.transformPage(wdsPage, recordType, queryOptions)
+  }
+}
+
+const recordType: string = 'mytype'
+
+const queryOptions: EntityQueryOptions = {
+  pageNumber: 2,
+  itemsPerPage: 50,
+  sortField: 'stringAttr',
+  sortDirection: 'desc',
+  snapshotName: '',
+  googleProject: '',
+  activeTextFilter: '',
+  filterOperator: ''
+}
+
+
+describe('WdsDataTableProvider', () => {
+  let getRecords
```

Comment:
some added type safety would be great here... try 
let getRecords: ReturnType<typeof Ajax>['WorkspaceDataService']['getRecords] = ...function block and body here...

and similar for the other methods being mocked here.

This should let you then set the props of WorkspaceDataService in the ajax mock further down as:
{
   getRecords: jest.fn().mockImplementation(getRecords),
   ...
}

One or two other patterns with good type safety are floating around as well... the functional instantiation of Ajax and the challenges in mocking it creates is a known issue.  But it may take a while before we can consider a pattern rework.

Reply:
exporting a WorkspaceDataServiceContract type from where WorkspaceDataService is defined would be helpful too, and shorten the type annotations here.  (But see other comment about renaming to WorkspaceData)

Reply:
an easier alternative would be to just add type decorations to these three vars:

`let getRecords: jest.MockedFunction<ReturnType<typeof WorkspaceDataService>['getRecords']>`

and similar for the other two.

My concern is that the soft cast near the end is not providing enough safety and is not alerting of type drift as well and early as this suggested change would.

`... as ReturnType<typeof Ajax>`

This should be a fairly easy drop-in... but if not, I won't block the PR for it.

Reply:
adding the type decorations to the `let` vars causes type-safety problems on the Ajax mock. I can make the Ajax mock work if I first cast it to unknown:
```typescript
asMockedFn(Ajax).mockImplementation(() => ({ WorkspaceData: { getRecords, deleteTable, downloadTsv } } as unknown as ReturnType<typeof Ajax>))
```
is that worth it?

Reply:
this is likely because you are only providing a small set of the total ajax contract.  First, for convenience, define 

`type AjaxContract = ReturnType<typeof Ajax>` 

Then casting the above to `as Partial<AjaxContract> as AjaxContract`